js/GraphFill.js

```diff
@@ -0,0 +1,141 @@
+export class Color {
+    static TRANSPARENT = new Color(0, 0, 0, 0);
+
+    static fromHex(literal) {
+        let c;
+
+        if (/^#([A-Fa-f0-9]{3}){1,2}$/.test(literal)) {
+            c = literal.substring(1).split('');
+
+            if (c.length === 3) {
+                c = [c[0], c[0], c[1], c[1], c[2], c[2]];
+            }
+
+            c = '0x' + c.join('');
+            return new Color((c >> 16) & 255, (c >> 8) & 255, c & 255);
+        }
+
+        throw new Error('Bad Hex');
+    }
+
+
+
+    #red;
+    #green;
+    #blue;
+    #alpha;
+
+    constructor(red, green, blue, alpha = 1) {
+        this.#red = red;
+        this.#green = green;
+        this.#blue = blue;
+        this.#alpha = alpha;
+    }
+
+
+
+    setAlpha(alpha) {
+        this.#alpha = alpha;
+        return this;
+    }
+
+    getAlpha() {
+        return this.#alpha;
+    }
+
+    clone() {
+        return new Color(
+            this.#red,
+            this.#green,
+            this.#blue,
+            this.#alpha
+        );
+    }
+
+    toString() {
+        return `rgba(${this.#red}, ${this.#green}, ${this.#blue}, ${this.#alpha})`;
+    }
+}
+
+
+
+export default class GraphFill {
+    #gradient;
+    #width;
+    #height;
+    /** @type {Color} */
+    #fill;
+    /** @type {Color} */
+    #border;
+    #colorUsed;
+
+
+
+    /**
+     * @param {Color} fill
+     * @param border
+     */
+    constructor(fill, border = undefined) {
+        this.#fill = fill;
+        this.#border = border ?? fill.clone().setAlpha(1);
```

Comment:
dont use ?? it breaks ingame overlay
```suggestion
        this.#border = border !== undefined ? border : fill.clone().setAlpha(1);
```

Reply:
true

Reply:
I wonder how because `a ?? b` is just syntax sugar for `a === undefined ? b : a`

Reply:
tosu's typescript doesnt really like it

Reply:
Null-ish coalescing operator was added in ES2020. Perhaps whatever JS engine tosu uses for its ingame overlay is just an older version.